ResearchKit/ActiveTasks/ORKGoNoGoViewController.m

```diff
@@ -0,0 +1,334 @@
+    }
+    if (lastReactionTime > 0)
+    {
+        NSString *reactionFormat = ORKLocalizedString(@"GONOGO_TASK_REACTION_FORMAT", nil);
+        NSString *reactionText = [NSString stringWithFormat:reactionFormat, lastReactionTime];
+        text = [text stringByAppendingString:reactionText];
+    }
+    
+    [self.activeStepView updateTitle:ORKLocalizedString(@"GONOGO_TASK_ACTIVE_STEP_TITLE", nil) text:text];
+}
+
+- (void)attemptDidFinish:(ORKResult *)result {
+    void (^completion)(void) = ^{
+        int successCount = 0;
+        for (ORKGoNoGoResult* res in _results) {
+            if (res.incorrect == NO) {
+                successCount++;
+            }
+        }
+        
+        if (successCount == [self gonogoTimeStep].numberOfAttempts) {
+            [self finish];
+        } else {
+            [self resetAfterDelay:2];
+        }
+    };
+    if(go) {
+        if (_validResult) {
+            [self indicateSuccess:completion result:result];
+        } else {
+            [self indicateFailure:completion result:result];
+        }
+    }
+    else {
+        if (_validResult) {
+            [self indicateFailure:completion result:result];
+        } else {
+            [self indicateSuccess:completion result:result];
+        }
+    }
+    
+    _validResult = NO;
+    _timedOut = NO;
+    [_stimulusTimer invalidate];
+    [_timeoutTimer invalidate];
+}
+
+- (void)indicateSuccess:(void(^)(void))completion result:(ORKResult *)result {
+    ORKGoNoGoResult *gonogoResult = [[ORKGoNoGoResult alloc] initWithIdentifier:self.step.identifier];
+    gonogoResult.timestamp = _stimulusTimestamp;
+    gonogoResult.timeToThreshold = [NSProcessInfo processInfo].systemUptime - _stimulusTimestamp;
+    if (result) {
+        gonogoResult.fileResult = (ORKFileResult *)result;
+    }
+    gonogoResult.go = go;
+    gonogoResult.incorrect = NO;
+    [_results addObject:gonogoResult];
+    
+    [_gonogoContentView startSuccessAnimationWithDuration:OutcomeAnimationDuration completion:completion];
+    AudioServicesPlaySystemSound([self gonogoTimeStep].successSound);
+}
+
+- (void)indicateFailure:(void(^)(void))completion result:(ORKResult *)result {
+    if (!_shouldIndicateFailure) {
+        return;
+    }
+    ORKGoNoGoResult *gonogoResult = [[ORKGoNoGoResult alloc] initWithIdentifier:self.step.identifier];
+    gonogoResult.timestamp = _stimulusTimestamp;
+    gonogoResult.timeToThreshold = [NSProcessInfo processInfo].systemUptime - _stimulusTimestamp;
+    if (result) {
+        gonogoResult.fileResult = (ORKFileResult *)result;
+    }
+    gonogoResult.go = go;
+    gonogoResult.incorrect = YES;
+    [_results addObject:gonogoResult];
+
+    
+    [_gonogoContentView startFailureAnimationWithDuration:OutcomeAnimationDuration completion:completion];
+    SystemSoundID sound = _timedOut ? [self gonogoTimeStep].timeoutSound : [self gonogoTimeStep].failureSound;
+    AudioServicesPlayAlertSound(sound);
+}
```

Comment:
You have a lot of code duplication here which can result in mistakes when something is updated later. Additionally, your completion block should always be the last field.

````
- (void)indicateResult:(ORKResult *)result incorrect:(BOOL)incorrect completion:(void(^)(void))completion {
    
    // Exit early if not recording the failure
    if (incorrect && !_shouldIndicateFailure) {
        return;
    }
    
    // Create a result
    ORKGoNoGoResult *gonogoResult = [[ORKGoNoGoResult alloc] initWithIdentifier:self.step.identifier];
    gonogoResult.timestamp = _stimulusTimestamp;
    gonogoResult.timeToThreshold = [NSProcessInfo processInfo].systemUptime - _stimulusTimestamp;
    if ([result isKindOfClass:[ORKFileResult class]]) {
        gonogoResult.fileResult = (ORKFileResult *)result;
    }
    gonogoResult.go = go;
    gonogoResult.incorrect = incorrect;
    [_results addObject:gonogoResult];
    
    // Start the animation and play the sound
    if (incorrect) {
        [_gonogoContentView startFailureAnimationWithDuration:OutcomeAnimationDuration completion:completion];
        SystemSoundID sound = _timedOut ? [self gonogoTimeStep].timeoutSound : [self gonogoTimeStep].failureSound;
        AudioServicesPlayAlertSound(sound);
    } else {
        [_gonogoContentView startSuccessAnimationWithDuration:OutcomeAnimationDuration completion:completion];
        AudioServicesPlaySystemSound([self gonogoTimeStep].successSound);
    }
}
````